core/dbt/adapters/base/impl.py

```diff
@@ -718,6 +718,24 @@ def list_relations(self, database: Optional[str], schema: str) -> List[BaseRelat
         # we can't build the relations cache because we don't have a
         # manifest so we can't run any operations.
         relations = self.list_relations_without_caching(schema_relation)
+
+        # if the cache is already populated, add this schema in
+        # otherwise, skip updating the cache and just ignore
+        if self.cache:
+            for relation in relations:
+                self.cache.add(relation)
+            if not relations:
+                # it's possible that there were no relations in some schemas. We want
+                # to insert the schemas we query into the cache's `.schemas` attribute
+                # so we can check it later
+
+                # Jeremy: what was the intent behind this inner loop?
+                # cache_update: Set[Tuple[Optional[str], Optional[str]]] = set()
+                # for relation in cache_schemas:
+                #     cache_update.add((database, schema))
+
+                self.cache.update_schemas(set((database, schema)))
```

Comment:
@jtcohen6 While pairing with @ChenyuLInx we weren't sure what this inner loop was doing, especially since `cache_schemas` isn't defined anywhere.

Reply:
@stu-k The cache is keyed on `database` + `schema`. When dbt does a cache lookup, it asks: Have I already cached this `database` + `schema` combo? If the combo isn't present as a key in the cache, it's a cache miss, and dbt needs to go run a query. If those keys are present, and no relations are found, then the assumption is that the schema is _empty_ (= missing from the database), rather than just missing from the cache.

If there are no `relations` returned by the query, we still want to record that, by inserting an _empty_ set for each `database` + `schema` pair. That way, the next time dbt wants to know if any relations are in that `database.schema`, it doesn't need to run the same query over again — the cache says, I already know there aren't any relations there.

Reply:
I understand that, but I do not understand the intent of your inner loop I've commented on here.
```python
cache_update: Set[Tuple[Optional[str], Optional[str]]] = set()
for relation in cache_schemas:
    cache_update.add((database, schema))
self.cache.update_schemas(cache_update)
```
What is this code trying to accomplish? It is difficult to determine because `cache_schemas` isn't defined in [your possible implementation on the original bug ticket](https://github.com/dbt-labs/dbt-core/issues/6240).